torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1643,58 +1841,108 @@ def _init_streams(self) -> None:
         """Initializes CUDA streams for overlapping data transfer and
         computation. This should only be called on the root FSDP instance."""
         assert self._is_root
-        if torch.cuda.is_available():
-            # Stream for all-gathering parameters.
-            self._streams["all_gather"] = torch.cuda.Stream()
-            # Stream for overlapping grad reduction with the backward pass.
-            self._streams["post_backward"] = torch.cuda.Stream()
-            # Stream to move main params to self.mixed_precision.param_dtype
-            # for forward pass.
-            if self._mixed_precision_enabled_for_params():
-                self._streams["mixed_precision_params"] = torch.cuda.Stream()
+        assert torch.cuda.is_available()
+        # Stream for all-gathering parameters.
+        self._streams["all_gather"] = torch.cuda.Stream()
+        # Stream for overlapping grad reduction with the backward pass.
+        self._streams["post_backward"] = torch.cuda.Stream()
+        # Stream for pre-all-gather copies (e.g. H2D or precision cast).
+        self._streams["pre_all_gather"] = torch.cuda.Stream()
 
     def _wait_for_previous_optim_step(self) -> None:
         """
         The root :class:`FullyShardedDataParallel` instance needs to
         synchronize with the default stream to ensure that the previous
         optimizer step is done.
         """
-        if not torch.cuda.is_available() or not self._is_root:
+        if not self._is_root:
             return
-        if self._mixed_precision_enabled_for_params():
-            self._streams["mixed_precision_params"].wait_stream(
-                torch.cuda.current_stream()
-            )
-        self._streams["all_gather"].wait_stream(torch.cuda.current_stream())
+        current_stream = torch.cuda.current_stream()
+        self._streams["all_gather"].wait_stream(current_stream)
+        # Having the pre-all-gather stream wait for the current stream even if
+        # we do not leverage the pre-all-gather stream is tolerable since this
+        # only runs once per iteration
+        self._streams["pre_all_gather"].wait_stream(current_stream)
```

Comment:
Looking forward at when we expose the original parameters (you can actually look at the PR now that it is open), we may want to do a storage check and writeback to the sharded flattened parameter in case any original parameters changed storage in the pre-unshard.

In that case, whether or not the logical pre-unshard is a no-op or not is no longer static. There is no way to know at the beginning of the iteration (when `_wait_for_previous_optim_step()` is called) whether or not the iteration will need to use the pre-unshard stream. The writeback must happen before casting to low precision because the original parameters are in full precision.

Given this, I feel that it is cleanest to always make this `wait_stream()` call. It should not have problematic overhead because (1) it is only called once per iteration and (2) when parameter mixed precision is enabled, we already call this.